models/user.go

```diff
@@ -1 +1,337 @@
 package models
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+
+	"golang.org/x/crypto/bcrypt"
+)
+
+const (
+	DEFAULT_LOCALE = "en"
+
+	USER_EMAIL_MAX_LENGTH = 32
+
+	USER_FIRST_NAME_MAX_RUNES = 64
+	USER_LAST_NAME_MAX_RUNES  = 64
+
+	USER_PASSWORD_MAX_LENGTH       = 32
+	USER_CONTACT_NUMBER_MAX_LENGTH = 20
+)
+
+// User contains the details about the user.
+type User struct {
+	ID                      string            `json:"id"`
+	CreatedAt               int64             `json:"created_at,omitempty"`
+	UpdatedAt               int64             `json:"updated_at,omitempty"`
+	DeletedAt               int64             `json:"deleted_at"`
+	Username                string            `json:"username"`
+	Password                string            `json:"password,omitempty"`
+	AuthData                *string           `json:"auth_data,omitempty"`
+	AuthService             string            `json:"auth_service"`
+	Email                   string            `json:"email"`
+	EmailVerified           bool              `json:"email_verified,omitempty"`
+	FirstName               string            `json:"first_name"`
+	LastName                string            `json:"last_name"`
+	ContactNumber           string            `json:"contact_number"`
+	FacebookUserID          string            `json:"facebook_user_id"`
+	GoogleUserID            string            `json:"google_user_id"`
+	Locale                  string            `json:"locale"`
+	Timezone                map[string]string `json:"timezone"`
+	TermsOfServiceId        string            `db:"-" json:"terms_of_service_id,omitempty"`
+	TermsOfServiceCreatedAt int64             `db:"-" json:"terms_of_service_created_at,omitempty"`
+}
+
+type UserPatch struct {
+	Username  *string           `json:"username"`
+	Password  *string           `json:"password,omitempty"`
+	FirstName *string           `json:"first_name"`
+	LastName  *string           `json:"last_name"`
+	Email     *string           `json:"email"`
+	Locale    *string           `json:"locale"`
+	Timezone  map[string]string `json:"timezone"`
+	RemoteId  *string           `json:"remote_id"`
+}
+
+type UserAuth struct {
```

Comment:
What do we plan to store in this AuthData and AuthService? Just trying to understand since we already store the google and facebook user id seperately.